resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsAllocator.scala

```diff
@@ -136,6 +151,10 @@ class ExecutorPodsAllocator(
 
   def isDeleted(executorId: String): Boolean = deletedExecutorIds.contains(executorId.toLong)
 
+  private[k8s] def stopApplication(exitCode: Int): Unit = {
+    sys.exit(exitCode)
+  }
```

Comment:
I'm not super sure on this logic here, generally speaking we want to do a more controlled shutdown than just a `sys.exit`, could we call into the shutdownhook like in YARN?

Or is there a reason why were using sys.ext here?

Reply:
I see the problem, and actually I don't know how to handle it properly ... 

I use `sys.exit` here just want to kill the process or driver pod, so that `ShutdownHookManager` will be invoked.

Reply:
Emm, I see a similar usage

https://github.com/apache/spark/blob/1d2159f888139e65c80db2003d521b0f684df83a/resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsSnapshotsStoreImpl.scala#L145-L147

Reply:
@holdenk sorry for so long late reply. I think the `sys.exit` should be fine here, because
1) According to docs of `Runtime#addShutdownHook`, `sys.exit` will trigger `ShutdownHookManager`, 
  > Registers a new virtual-machine shutdown hook.
The Java virtual machine shuts down in response to two kinds of events:
The program exits normally, when the last non-daemon thread exits or when the exit (equivalently, System.exit) method is invoked, or
The virtual machine is terminated in response to a user interrupt, such as typing ^C, or a system-wide event, such as user logoff or system shutdown.

2) I read the code of YARN `ApplicationMaster`, the extra logic consists of
  - clean staging dir
  - resubmit if `exitCode != 0` and max attempts is unreached, or call `YarnRMClient.unregister` to unregister application.

IMO we don't need to do 2) in K8s, simply call `sys.exit(exitCode)` to propagate the exit code to Pod state should be fine.